suitcase/utils/__init__.py

```diff
@@ -1,25 +1,219 @@
+    def __exit__(*except_detail):
+        pass  # this avoids closing the file handle too early.
+
+
+class MemoryBuffersWrapper:
+    """
+    A class that manages multiple StringIO and/or BytesIO instances.
+
+    This design is inspired by Python's zipfile and tarfile libraries.
+
+    This has a special buffers attribute which can be used to retrieve
+    buffers created.
+    """
+    def __init__(self):
+        self._reserved_names = set()
+        self._artifacts = collections.defaultdict(list)
+        self.buffers = {}  # maps postfixes to buffer objects
+
+    @property
+    def artifacts(self):
+        return dict(self. _artifacts)
+
+    def reserve_name(self, label, postfix):
+        """
+        Ask the wrapper for a filepath.
+
+        An external library that needs a filepath (not a handle)
+        may use this instead of the ``open`` method.
+
+        Parameters
+        -------
+        label : string
+            partial file name (i.e. stream name)
+        postfix : string
+            relative file path and filename
+
+        Returns
+        ----------
+        filepath : Path
+         """
+        raise SuitcaseUtilsTypeError(
+            "MemoryBuffersWrapper is incompatible with exporters that require "
+            "explicit filenames.")
+
+    def open(self, label, postfix, mode, encoding=None, errors=None):
+        """
+        Request a file handle.
+
+        Like the built-in open function, this may be used as a context manager.
+
+        Parameters
+        -------
+        label : string
+            partial file name (i.e. stream name)
+        postfix : string
+            relative file path and filename
+        mode : {'x', 'xt', xb'}
+            'x' or 'xt' for text, 'xb' for binary
+        encoding : string or None
+            Not used. Accepted for compatibility with built-in open().
+        errors : string or None
+            Not used. Accepted for compatibility with built-in open().
+
+       Returns
+       -------
+       file : handle
+        """
+        # Of course, in-memory buffers have no filepath, but we still expect
+        # postfix to be a thing that looks like a relative filepath, and we use
+        # it as a unique identifier for a given buffer.
+        if Path(postfix).is_absolute():
+            raise SuitcaseUtilsValueError(
+                f"{postfix} must be structured like a relative file path.")
+        name = Path(postfix).expanduser().resolve()
+        if name in self._reserved_names:
+            raise SuitcaseUtilsValueError(
+                f"The postfix {postfix!r} has already been used.")
+        self._reserved_names.add(name)
+        self._artifacts[label].append(name)
+        if mode in ('x', 'xt'):
+            buffer = PersistentStringIO()
+        elif mode == 'xb':
+            buffer = PersistentBytesIO()
+        else:
+            raise SuitcaseUtilsModeError(
+                f'the mode passed to MemoryBuffersWrapper.open is {mode} but '
```

Comment:
It's more conventional to use double quotes around strings with text "for humans" so that if a contraction or other apostrophe slips into the text, it works with extra thought / escaping. Strings for computers are usually written with single quotes like `'xt'`. Also, if `f` is used at all, use it on every line.